A step toward modular CSS (and so more modular components).

ahdinosaur opened the pull request:
%CcTHlP33vXa9mA/lf8dxkUrZaqXA1DcXyZeRVHNb7Z4=.sha256

[`git.scuttlebot.io` link](http://git.scuttlebot.io/%25CcTHlP33vXa9mA%2Flf8dxkUrZaqXA1DcXyZeRVHNb7Z4%3D.sha256)

dominictarr commented:
is this ready to go? can you describe the changes?

evbogue commented:
I tested this branch in the lite client, and I'm getting `setImmediate is not defined`. I'm also concerned that the fs.readFileSync will also not work in the lite client.

dominictarr commented:
If we built it `brfs` we can make `fs.readFileSync` however I would prefer to make mcss compile to javascript so it can be loaded with require, because that will some plans I have down the road a bit easier.

ahdinosaur commented:
yes, this is more or less ready.

> each module that wants to export styles can either export `css` (a function that returns a css string) or `mcss` (a function that returns a `micro-css` string). these are then merged together in a `styles` module, which is then included to the document head in the `app` module.
>
> for themers, it's only necessary to override the `styles` module and change how the styles are merged together, either with a provided module-override-map or a full complete stylesheet (to match class names in modules).
>
> also chose to keep the current stylesheet the same, so this can be a small change to get the ball rolling in this direction.

also:

- moved backlinks code into a separate module
- changed timestamp export to not be `message_meta`
- updated about messages (added rendering of `description`), also add mini renderer
- add `message_author` which renders avatar, name, and timestamp, also add size option when you want to minify (not sure about `size`, just trying to avoid a `mini` boolean option when there are many possible sizes)

and then probably the most worth discussion: changed the layout of message. originally @mmckegg, @mixmix, and i split this into header, content, and footer, where header contained the author (avatar, name, timestamp) and meta in a flex row, and footer contained actions and backlinks. but then later that night i changed it so everything was in the same container, with author float left, meta float right, then content, actions, and backlinks.

it looks like this:

![patchbay message layout](https://cloud.githubusercontent.com/assets/719605/21465218/6ea82940-c991-11e6-897d-a778885379ff.png)

conceptually, it's this:

```txt
------------------------------------------------------
|  author  |           content            |  meta     |
|----------                               ------------|
|                                                     |
|                                                     |
|-----------------------------------------------------|
|                      actions                        |
|-----------------------------------------------------|
|                     backlinks                       |
|-----------------------------------------------------|
```

however now i wonder if i should make the top a separate module that you can add to (like adding "re: blah blah blah" out of the content into that section). same goes with maybe splitting the bottom into three sections, or something.

thoughts?

mixmix commented:
initial thought before I look at code: that box model is going to break bad for long names - you can already see it making the content width variable for the different names in the example. Might be avoidable with styling but not sure that should be the solution?

mixmix commented on modules_basic/index.js:
```diff
@@ -12,6 +12,8 @@ module.exports = {
+  "message-backlinks.js":  require('./message-backlinks.js'),
   "message-link.js":  require('./message-link.js'),
   "message-name.js":  require('./message-name.js'),
   "message.js":  require('./message.js'),
```
would prefer message to be above the message partials ... 
notice in this is starting to feel like it could like a folder....

ahdinosaur replied:
yeah  it _is_ an alphabetical folder listing. :wink: so could replace with `bulk-require` but another pull request.

mixmix commented on modules_basic/message-author.js:
```diff
@@ -0,0 +1,39 @@
+    var { value } = msg
+    var { author } = value
+    return h('MessageAuthor', {
+      className: size === 0 ? '-mini' : ''
```
is this your opinion? 
are there a range of sizes here ? If not would prefer a boolean opts, like `isMini`

ahdinosaur replied:
yeah, "add size option when you want to minify (not sure about size, just trying to avoid a mini boolean option when there are many possible sizes)". boolean opts like `isMini` are the worst, when really we want either enumerated sizes or a range of sizes (i like the range between 0 and 1). probably enumerated is better, will change `size` to a string.

mixmix replied:
not sure what 'change size to string' means ... you mean `size: mini` ? I think that might make sense

mixmix commented on modules_basic/message-author.mcss:
```diff
@@ -0,0 +1,12 @@
+  flex-direction: column
+
+  -mini {
+    flex-direction: row
```
I'm really like this modification pattern

mixmix commented on modules_basic/message.js:
```diff
@@ -1,122 +1,86 @@
 
-    var el = api.message_content(msg)
-    if(!el) return mini(msg, message_content_mini_fallback(msg))
+    var content = api.message_content(msg)
```
second declaration of variable (remove var)

mixmix commented on modules_basic/message.js:
```diff
@@ -1,122 +1,86 @@
-    msg.setAttribute('tabindex', '0')
-
-    return msg
+  function mini(msg, el) {
```
this pattern feels a bit weird. I had to read it a couple of times, this feels like a template.
I wonder if it would be good to have a `miniTemplate` and a `standardTemplate` both extracted so that we could see this pattern more clearly?

mixmix commented on modules_basic/reply.js:
```diff
@@ -0,0 +1,9 @@
+
+exports.gives = 'message_action'
+
+exports.create = function () {
```
other creates take in api even though they don't use them. I think we should be consistent to help other people see the pattern

mixmix replied:
I wonder about an empty needs as well, if we're being super explicit

ahdinosaur replied:
oh, i thought i saw other modules that didn't need anything not receiving api. i'm happy either way.

mixmix commented on modules_basic/timestamp.js:
```diff
@@ -1,28 +1,47 @@
-    return el
-  }
-
   setInterval(function () {
```
right, this is probably where mutation is being introduced - and needs setImmediate

mixmix commented on modules_basic/timestamp.js:
```diff
@@ -1,28 +1,47 @@
-
   setInterval(function () {
-    var els = [].slice.call(document.querySelectorAll('.timestamp'))
+    var els = [].slice.call(document.querySelectorAll('.Timestamp'))
```
oh this makes me feel weird / sad

mixmix replied:
it feels like it breaks some rule... like effects outside of the depject flow.
something to watch

mmckegg replied:
Yeah, I really dislike this. But mostly just because it is not very mutanty - looks a lot more like jquery.

There are much better ways to do this. For example with a hook. Could still be depjected.

ahdinosaur replied:
agree, keen to learn how we could do this better.

mixmix commented on modules_core/app.js:
```diff
@@ -1,12 +1,18 @@
   create: function (api) {
     return function () {
-      document.head.appendChild(h('style', require('../style.css.json')))
+      process.nextTick(function () {
```
I'm a noob to this, but there might be a better way to do this if we have setImmediate around ? 
Might be worth checking how Matt did this if you haven't already

ahdinosaur replied:
what do you mean? Matt also used `insert-css`, the `process.nextTick` is because of how depject is currently written. see https://github.com/dominictarr/depject/pull/2 for a pull request that would remove the need for `process.nextTick`.

mixmix replied:
ok, I've not looked at the details. trust your knowledge on this

mixmix commented on modules_core/styles.js:
```diff
@@ -0,0 +1,34 @@
+module.exports = {
+  needs: {
+    mcss: 'map',
+    css: 'map'
```
guess  I wonder whether we should add css as an option - I think if we're doing mcss, lets stick to that and be pushed around by the opinions it has. I wouldn't want to see people adding piecemeal css just _because it's quicker_

ahdinosaur replied:
you sure? it seems wrong to me to not give the option of plain css.

mixmix replied:
I dunno, I was trying a strong stance. I reckon we could try a strong stance (maybe with the css lines commented out to start with)
Feels like a case where it might be good and bringing it back from css to mcss might be a pain

ahdinosaur commented on package.json:
```diff
@@ -20,9 +22,11 @@
     "is-visible": "^2.0.4",
     "kvgraph": "^0.1.0",
     "map-filter-reduce": "^3.0.1",
+    "micro-css": "^0.6.2",
```
i've only been using chrome or electron, which has `setImmediate`. i think it's a dependency of `mutant`.

mixmix replied:
yeah it is. can we add the setimmediate package ?

ahdinosaur commented:
made changes, @evbogue can you see if this works for you now?

evbogue commented:
The lite client builds now.

mixmix commented:
hey @ahdinosaur did you sync this with ssb recently ? 

I tried making some changes to padding in response to ahhh arj's padding feedback and got some good looking stuff going on BUT ... the flex with longer content looks like it's not going to work : 

here's a tame but kinda broke example: 
![selection_146](https://cloud.githubusercontent.com/assets/2665886/21514978/3d374ca4-cd30-11e6-9570-1f5df13c6568.png)

here's a well broken one (in terms of readability) : 
![selection_147](https://cloud.githubusercontent.com/assets/2665886/21514981/4bf4418e-cd30-11e6-8543-1bbba0c1dd8f.png)

ahdinosaur commented:
@mixmix that seems to be expected behavior given the box model, no?

that's the inspiration for my comment:

> however now i wonder if i should make the top a separate module that you can add to (like adding "re: blah blah blah" out of the content into that section). same goes with maybe splitting the bottom into three sections, or something.

so the box model becomes

```txt
------------------------------------------------------
|  author  |          frontlinks          |  meta     |
|---------------------------------------- ------------|
|                      content                        |
|-----------------------------------------------------|
|                      actions                        |
|-----------------------------------------------------|
|                     backlinks                       |
|-----------------------------------------------------|

````

would this address the issue you have in mind?

mixmix commented:
the styling in those screenshots was in this branch  `about_refactor_styling` on ssb

mixmix commented:
@ahdinosaur yeah I reckon it could. 

this feels like that structural opinionated edge. could also be : 

```
|-----------------------------------------------------|
|  author  |          frontlinks          |  meta     |
|          |----------------------------- ------------|
|          |           content                        |
|          |------------------------------------------|
|          |           actions                        |
|          |------------------------------------------|
|          |          backlinks                       |
|----------|------------------------------------------|
```

mixmix commented:
I was wondering if the box model could be 

```
- Message (flex flex-direction: row)
  - Author
  - Other (flex flex-direction: col)
    - Info (flex flex-direction: row)
      - FrontLinks
      - Meta
    - Content
    - Actions
    - Backlinks
```

That way we people could flick the flex direction to column and it would fall back to what you had ?  (This is a naive guess, might not work like that with flex ...)

mixmix commented:
boom this : 

```
 ------------------------------------------------------
| ---------- ------------------------------------------ |
||  author  | ----------------------------- ---------- ||
||          ||         frontlinks          |  meta    |||
||          | ----------------------------- ---------- ||
||          |------------------------------------------||
||          |           content                        ||
||          |------------------------------------------||
||          |           actions                        ||
||          |------------------------------------------||
||          |          backlinks                       ||
| ---------- ------------------------------------------ |
 ------------------------------------------------------- 
```

ahdinosaur commented:
yeah, i was trying to avoid nesting as it intuitively feels more rigid, but maybe that's not true. i also wasn't using flexbox, because i couldn't figure out how to make it work without adding container elements (as we had originally at Art~Hack).

anyways, i gave a go at [implementing my idea](https://github.com/ssbc/patchbay/commit/b3f540ed553d092084367d6ea9ed825fcb020048) (except `frontlinks` became `title`): [`dinostyle-refactor`](https://github.com/ssbc/patchbay/compare/dinostyle-refactor)

![snapshot](https://cloud.githubusercontent.com/assets/719605/21520864/a25ed274-cd5c-11e6-9690-e57dd3c78146.png)

ahdinosaur commented:
what i mean by "not nested" is [this](https://github.com/ssbc/patchbay/blob/b3f540ed553d092084367d6ea9ed825fcb020048/modules_basic/message.js#L38-L50):

```txt
- Message
  - Author (float: left)
  - title (float: left)
  - meta (float: right)
  - content (clear: both)
  - action
  - Backlinks
```

it felt like minimum viable opinion.

dominictarr commented:
what if the avatar image floats, but the name is on the top? I think that might be the best of both?

mixmix commented:
@ahdinosaur it feels like not that MVP, in that the nice flat strucutre doesn't feel like it will allow for much useful styling (to me). Maybe it's possible with CSS wizardry, but I feel less excited by that.

Maybe you could tell us a little more about your motivation for such a flat box-model structure ?

mixmix commented:
@ahdinosaur here's a possible solution with styling ... if negative margin is ok.

![selection_148](https://cloud.githubusercontent.com/assets/2665886/21532892/dd32a9fe-cdb9-11e6-8857-e494a42ed6ac.png)

It breaks the mini messages but I think that's probably fixable.
pushing this to github on branch `mixstyle-refactor`

ahdinosaur commented:
@dominictarr can you explain more why you want the name on top?

as opposed to:

```txt
[image] | title, message meta
 name   | ------------------
 time   | content
 ---------------------------
 backlinks, actions
```

@mixmix i like that, can you push up your branch so i can see what you did?

evbogue commented:
It could be that the position of the date _is_ opinionated. Which may mean we need to leave that choice up to the developer of an individual Bay-based sbot client.

mixmix commented:
@ahdinosaur pushed ! 

got something working with mini messages. 
I notice the mini message might feel better if it was MessageMini rather than Message -mini

it feels like I was over-riding too much maybe... something to watch 

![selection_149](https://cloud.githubusercontent.com/assets/2665886/21533693/4bf86790-cdc2-11e6-8ef7-ad9418fc75d8.png)

ahdinosaur commented:
thanks @mixmix, that looks great. :smiley_cat: i feel -mini is working exactly as intended. _shrug_

@evbogue yep, to override this you'd want to override `MessageAuthor` to not include it and use the timestamp module somewhere else, like export as a meta.

arj03 commented:
Really digging the mixmix refactor branch. Only wierd thing I noticed besides long names, is the private messages that looks a bit wierd.

mixmix commented:
new pull request opened to merge changes into this branch

arj03 commented:
Seems like the latest version doesn't run well with lite client

mixmix commented:
thanks for the heads up @arj03 . 
How are you running the lite client (I've obviously not looked at that side of the project), and what browser are you using? I trust you `npm install`ed after pulling, there are a couple of new modules?

arj03 commented:
I'm getting these two errors:

insert-css: You need to provide a CSS string. Usage: insertCss(cssString[, options]).
Error: insert-css: You need to provide a CSS string. Usage: insertCss(cssString[, options]).
    at insertCss (http://localhost:8080/build/index.html:12039:15)
    at http://localhost:8080/build/index.html:2341:9
    at Item.run (http://localhost:8080/build/index.html:6399:14)
    at drainQueue (http://localhost:8080/build/index.html:6369:42)
    at cleanUpNextTick (http://localhost:8080/build/index.html:6352:9)

fs is not defined
ReferenceError: fs is not defined
    at http://localhost:8080/build/index.html:43:12
    at http://localhost:8080/build/index.html:8938:19
    at Array.map (native)
    at Object.mcss (http://localhost:8080/build/index.html:8937:19)
    at http://localhost:8080/build/index.html:2793:22
    at Item.run (http://localhost:8080/build/index.html:6399:14)
    at drainQueue (http://localhost:8080/build/index.html:6369:42)

arj03 commented:
@mixmix much better thanks 👍
